### PR TITLE
Added .games command

### LINKED ISSUE #278: [Feature Request] Horror Game Recommendations
It's that spooky time of the year, and there's no better time to turn off the lights and cower in fear in front of a screen! To also keep up with the gaming spirit, what game should you play? Ask Spookybot!

## Implementation details
- [ ] Return an embed containing information on a random popular horror game

## Information in the embed
- Summary of the game
- Developer of the game
- Publisher of the game
- Rating
- Release Date
- IGN score

### LINKED ISSUE #354: [Feature Request] .games Command
## Description

There should `.games` command where you must provide genre to get random games from this genre (like my added `.movies` command) and also have following subcommands:
- `top <amount>` - Most popular games
- `company` - Get random game company

## Reasoning

I'd like to see this in SeasonalBot, and there is already something like this in Halloween Season, so this mean other peoples want this too.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [ ] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by ks129**
## Relevant Issues

Closes #354
Closes #278 

## Description

I added `.games <genre> <amount (default 5)>` command + all it's subcommands:
- `.games top <amount, default 10>` - Get current top games
- `.games search <search string>` - Search games by title
- `.games genres` - Get all current available genres
- `.games companies <amount default 5>` - Get random game companies

I used IGDB API for this.

## Reasoning

I added this due I lately added `.movies` command and current game recommendation command is pretty limited.

## Screenshots

Screenshot 2020-02-24 at 18 59 48
Screenshot 2020-02-24 at 19 00 11
Screenshot 2020-02-24 at 19 00 43
Screenshot 2020-02-24 at 19 00 57

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

File: bot/seasons/evergreen/game.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+            pages.append(await self.create_page(game))
+        return pages
+
+    async def create_page(self, data: Dict[str, Any]) -> Tuple[str, str]:
+        """Create content of Game Page."""
+        # Create page content variable, what will be returned
+        page = ""
+
+        # Get keys to one variable. Used for code style
+        keys = data.keys()
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]: Why is this needed?  The use of it seems inconsistent with for instance L236 and L224. Might as well use data in those places?

File: bot/seasons/evergreen/game.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+            data = await resp.json()
+
+        # Loop over games, format them to good format, make line and append this to total lines
+        for game in data:
+            line = ""
+
+            # Add games name and rating, also attach URL to title
+            line += f"**[{game['name']}]({game['url']})**\n"
+            line += f"""{'{0:.2f}'.format(game['total_rating'] if 'total_rating' in game.keys() else 0)}/100 :star: (by {
+            game['total_rating_count'] if 'total_rating_count' in game.keys() else '?'} users)"""
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]:
```suggestion
            game['total_rating_count'] if 'total_rating_count' in game else '?'} users)"""
```

Comment (tagptroll1) [reply bot/seasons/evergreen/game.py]: You get the idea

File: bot/seasons/evergreen/game.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+    async def get_company_pages(self, data: List[Dict[str, Any]]) -> List[Tuple[str, str]]:
+        """Get all good formatted pages about Game Companies."""
+        pages = []
+
+        # Loop over companies, add them to pages listing and return pages
+        for co in data:
+            pages.append(await self.create_company_page(co))
+
+        return pages
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]: Couldn't this be a list comprehension?

File: bot/seasons/evergreen/game.py [hunk 4]
```diff
@@ -198,42 +208,38 @@ def __init__(self, bot: Bot):
     async def get_pages(self, data: List[Dict[str, Any]]) -> List[Tuple[str, str]]:
         """Generate all game pages, do additional requests to IGDB API."""
         pages = []
-        for game in data:
-            pages.append(await self.create_page(game))
+        [pages.append(await self.create_page(game)) for game in data]
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]: What about 
```suggestion
        pages = [await self.create_page(game) for game in data]
```
Doesn't that work?

File: bot/seasons/evergreen/game.py [hunk 5]
```diff
@@ -198,42 +208,38 @@ def __init__(self, bot: Bot):
     async def get_pages(self, data: List[Dict[str, Any]]) -> List[Tuple[str, str]]:
         """Generate all game pages, do additional requests to IGDB API."""
         pages = []
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]:
```suggestion
```
with suggestion below

File: bot/seasons/evergreen/game.py [hunk 6]
```diff
@@ -296,8 +302,7 @@ def __init__(self, bot: Bot):
...
 
         # Loop over companies, add them to pages listing and return pages
-        for co in data:
-            pages.append(await self.create_company_page(co))
+        [pages.append(await self.create_company_page(co)) for co in data]
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]: See comment on L211

File: bot/seasons/evergreen/game.py [hunk 7]
```diff
@@ -198,42 +208,38 @@ def __init__(self, bot: Bot):
...
         page += f"**Platforms:** "
-        page += f"{', '.join(pf['name'] for pf in data['platforms']) if 'platforms' in keys else '?'}\n"
+        page += f"{', '.join(pf['name'] for pf in data['platforms']) if 'platforms' in data else '?'}\n"
 
-        page += f"**Status:** {GameStatus(data['status']).name if 'status' in keys else '?'}\n"
+        page += f"**Status:** {GameStatus(data['status']).name if 'status' in data else '?'}\n"
 
-        if 'age_ratings' in keys:
+        if 'age_ratings' in data:
             rating = f"""{', '.join(AgeRatingCategories(age['category']).name + ' ' + AgeRatings(age['rating']).name
```
Comment (tagptroll1) [inline bot/seasons/evergreen/game.py]: Why does this need an f-string wrapped around it?  join returns a string. I'd use an f-string inside the join, but not around it.
```
rating = ', '.join(
    f'{AgeRatingCategories(age["category"]).name} {AgeRatings(age["rating"]).name}'
    for age
    in date['age_ratings']
)
```

File: bot/seasons/evergreen/game.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+            elif genre == "Turn-based strategy (tbs)":
+                self.genres["Turn-based-strategy"] = genres[genre]
+                self.genres["Tbs"] = genres[genre]
+            elif genre == "Real time strategy (rts)":
+                self.genres["Real-time-strategy"] = genres[genre]
+                self.genres["Rts"] = genres[genre]
+            elif genre == "Hack and slash/beat 'em up":
+                self.genres["Hack-and-slash"] = genres[genre]
+            else:
+                self.genres[genre] = genres[genre]
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: Looking over this is very confusing. Correct me if I'm wrong, but you're creating aliases for long genre names.

It would be a lot more maintainable to handle the aliasing logic in a generic loop, test it to make sure it works properly, and then simply define the aliases on the side.
```python
aliases = {
    "Role-playing (rpg)": ["Role-playing", "Rpg"],
}

for original, aliases in aliases.items():
    for alias in aliases:
        genres[alias] = original[original]
```
In this particular case, both aliases are case-less sub-strings of the original. There's room for improvement here in terms of how much the cog could aid the user with look-ups.

File: bot/seasons/evergreen/game.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        pages = await self.get_pages(games)
+        await ImagePaginator.paginate(pages, ctx, Embed(title=f'Top {amount} Games'))
+
+    @games.command(name='genres', aliases=['genre', 'g'])
+    async def genres(self, ctx: Context) -> None:
+        """Get all available genres."""
+        await ctx.send(f"Currently available genres: {', '.join(f'`{genre}`' for genre in self.genres)}")
+
+    @games.command(name='search', aliases=['s'])
+    async def search(self, ctx: Context, *, search: str) -> None:
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: Consider using a better name for the param, such as `search_term`.

File: bot/seasons/evergreen/game.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        Get random Game Companies companies from IGDB API.
+
+        Support amount parameter. Max is 25, min is 1.
+        """
+        if amount > 25:
+            await ctx.send("You can't get more than 25 companies at once.")
+            return
+        elif amount < 1:
+            await ctx.send("You can't get less than 1 company.")
+            return
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: I think just checking if `1 <= amount <= 25` would be sufficient. You can then explain to the user that `amount` escapes the allowed range, and show what the range is.

File: bot/seasons/evergreen/game.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        await ImagePaginator.paginate(pages, ctx, Embed(title='Random Game Companies'))
+
+    async def get_games_list(self,
+                             client: ClientSession,
+                             limit: int,
+                             genre: str = None,
+                             sort: str = None,
+                             additional_body: str = "",
+                             offset: int = 0) \
+            -> List[Dict[str, Any]]:
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: The prevalent style for long function signatures is as follows:
```python
def my_function(
    a: type,
    b: type = default
) -> return_type:
```
Please try to avoid line continuation with `\`. It is *very* rarely necessary.

Additionally:
- Does `client` need to be a param? You always pass in `self.http_session`. This function can simply grab the session from `self`.
- I'd prefer using `amount` rather than `limit`, especially if the parameter isn't explained anywhere. When calling the function, it is referred to as `amount`.
- What are allowed values for `sort`? What does it do?
- What am I supposed to pass into `additional_body`?
- What is `offset`?

A docstring of `Get Games List from IGDB API.` may not cut it for this function signature.

Regarding the type annotations - if a param may be of type `T` or `None`, annotate it as `Optional[T]`. Some of these string params default to None, whereas some default to an empty string. It's up to you to choose which one makes more sense: if the query can be safely constructed with empty strings, you can save yourself a conditional. If you'll have to check whether a non-default param was provided, using `None` to denote "no param was passed" in a conditional may look cleaner.

File: bot/seasons/evergreen/game.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        body = "fields cover.image_id, first_release_date, total_rating, name, storyline, url, platforms.name, "
+        body += "status, involved_companies.company.name, summary, age_ratings.category, age_ratings.rating, "
+        body += "total_rating_count;\n"
+
+        body += f"sort {sort};\n" if sort else ''
+        body += f"offset {offset};\n"
+        body += f"limit {limit};\n"
+
+        body += f"where genres = ({genre});" if genre else ''
+        body += additional_body
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: I find the repeated concatenation unnecessary. A string that won't fit into one line can be constructed as such:
```python
body = (
    "a very long "
    "message."
)
```
This won't distract from the content. Using f-strings remains allowed. If the conditionals are necessary, it may be better to assign the desired values to variables with descriptive names, and then constructing the string from them. Since this is an API query, you may want to consider using a template string at the top of the module, and then simply passing in these optional parameters to it as necessary. It's probably ok to keep it here though.

File: bot/seasons/evergreen/game.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+            page += f"**Age Ratings:** {rating}\n"
+
+        if 'involved_companies' in data:
+            companies = ', '.join(co['company']['name'] for co in data['involved_companies'])
+        else:
+            companies = "?"
+        page += f"**Made by:** {companies}\n"
+
+        page += "\n"
+        page += data['storyline'] if 'storyline' in data else ''
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: This is beyond unreadable. Please consider using a template string and calling `.format` on it. Intermediate parts should be built separately and then joined together.

See my other comment regarding repeated concatenations.

File: bot/seasons/evergreen/game.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+        # Define request body of IGDB API request and do request
+        body = f"""fields name, url, storyline, total_rating, total_rating_count;
+search "{search}";
+limit 50;"""
+
+        async with client.get(
+                url=BASE_URL + "games",
+                data=body,
+                headers=HEADERS) as resp:
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: On line 297, you style a similar expression in a preferred manner:
```python
async with client.get(
    url=BASE_URL + "companies",
    data=body,
    headers=HEADERS
) as resp:
```
Without getting into a discussion of which one is more "correct", it's good to be consistent.

By the way, this whole thing will comfortably fit into one line:
```python
async with client.get(url=BASE_URL + "companies", data=body, headers=HEADERS) as resp:
```

File: bot/seasons/evergreen/game.py [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+            lines.append(line)
+
+        return lines
+
+    async def get_companies_list(self, client: ClientSession, limit: int, offset: int = 0) -> List[Dict[str, Any]]:
+        """Get random Game Companies from IGDB API."""
+        # Create request body, define included fields, limit and offset, do request
+        body = f"""fields name, url, start_date, logo.image_id, developed.name, published.name, description;
+limit {limit};
+offset {offset};"""
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: Please see my other comments regarding long string construction. Is it necessary that the query contains newlines?

File: bot/seasons/evergreen/game.py [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
+class Games(Cog):
+    """Games Cog contains commands that collect data from IGDB."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.http_session: ClientSession = bot.http_session
+
+        # Initialize genres
+        asyncio.get_event_loop().create_task(self._get_genres())
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: You can simply use the bot's own event loop: `bot.loop`. It's good to assign the returned task to an instance variable. That way, you can check on it later, if necessary for debug or other functionality, e.g. to check whether it has already finished.

File: bot/seasons/evergreen/game.py [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+                self.genres["Rts"] = genres[genre]
+            elif genre == "Hack and slash/beat 'em up":
+                self.genres["Hack-and-slash"] = genres[genre]
+            else:
+                self.genres[genre] = genres[genre]
+
+    @group(name='games', aliases=['game'], invoke_without_command=True)
+    async def games(self, ctx: Context, genre: str = "", amount: int = 5) -> None:
+        """
+        Get random game(s) by genre from IGDB. Use .movies genres command to get all available genres.
```
Comment (kwzrd) [inline bot/seasons/evergreen/game.py]: You advise the user here to call `.movies`. I assume this was copy-pasted from the `Movie` cog.

File: bot/seasons/evergreen/game.py [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+        self.bot = bot
+        self.http_session: ClientSession = bot.http_session
+
+        # Initialize genres
+        asyncio.get_event_loop().create_task(self._get_genres())
+
+    async def _get_genres(self) -> None:
+        """Create genres variable for games command."""
+        body = "fields name; limit 100;"
+        async with self.http_session.get(BASE_URL + "genres", data=body, headers=HEADERS) as resp:
```
Comment (sco1) [inline bot/seasons/evergreen/game.py]: Let's use an fstring here

```suggestion
        async with self.http_session.get(f"{BASE_URL}genres", data=body, headers=HEADERS) as resp:
```

It's probably more readable to strip the trailing `/` from `BASE_URL` so these concatenations would end up being `f"{BASE_URL}/genres"` instead.

File: bot/seasons/evergreen/game.py [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+        elif amount < 1:
+            await ctx.send("You can't get less than 1 game.")
+            return
+
+        # Get games listing, if genre don't exist, show help.
+        try:
+            games = await self.get_games_list(self.http_session, amount, self.genres[genre],
+                                              offset=random.randint(0, 150))
+        except KeyError:
+            await ctx.send_help('games')
```
Comment (sco1) [inline bot/seasons/evergreen/game.py]: I don't think just sending the command's help here is useful. There should be feedback to the user that there's something wrong with the genre they provided.

File: bot/seasons/evergreen/game.py [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+            data = await resp.json()
+
+        # Loop over games, format them to good format, make line and append this to total lines
+        for game in data:
+            line = ""
+
+            # Add games name and rating, also attach URL to title
+            line += f"**[{game['name']}]({game['url']})**\n"
+            line += f"""{'{0:.2f}'.format(game['total_rating'] if 'total_rating' in game.keys() else 0)}/100 :star: (by {
+            game['total_rating_count'] if 'total_rating_count' in game else '?'} users)"""
```
Comment (sco1) [inline bot/seasons/evergreen/game.py]: Having to mix so many things and use a multiline string on top of it is generally an indication that this should be broken up into discrete chunks.

**Comment by ks129** [comment]
@sco1 @kwzrd After 6 hours of work, this is done.

File: bot/seasons/evergreen/game.py [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+                             offset: int = 0
+                             ) -> List[Dict[str, Any]]:
+        """
+        Get list of games from IGDB API by parameters that is provided.
+
+        Amount param show how much movies this get, genre is genre ID and at least one genre in game must this when
+        provided. Sort is sorting by specific field and direction, ex. total_rating desc/asc (total_rating is field,
+        desc/asc is direction). Additional_body is field where you can pass extra search parameters. Offset show start
+        position in API.
+        """
```
Comment (eivl) [inline bot/seasons/evergreen/game.py]: The comment refers to movies. should be changed to games.

File: bot/seasons/evergreen/game.py [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+
+        # Check for amounts, max is 25 and min 1
+        if not 1 <= amount <= 25:
+            await ctx.send("Your provided amount is out of range. Our minimum is 1 and maximum 25.")
+            return
+
+        # Get games listing, if genre don't exist, show error message with possibilities.
+        try:
+            games = await self.get_games_list(amount, self.genres[genre],
+                                              offset=random.randint(0, 150))
```
Comment (eivl) [inline bot/seasons/evergreen/game.py]: I do not find the offset, in this case, to be self-documented. Had to track down the `get_games_list` function to understand what it does. Maybe a clarification of why the offset is picked at random could be written as a comment?

File: bot/seasons/evergreen/game.py [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+        """
+        Get random Game Companies companies from IGDB API.
+
+        Support amount parameter. Max is 25, min is 1.
+        """
+        if not 1 <= amount <= 25:
+            await ctx.send("Your provided amount is out of range. Our minimum is 1 and maximum 25.")
+            return
+
+        companies = await self.get_companies_list(amount, random.randint(0, 150))
```
Comment (eivl) [inline bot/seasons/evergreen/game.py]: I would love keyword arguments so that I would know that this was a random offset. Same as the other ones
It's not self-documented so I have to look up `get_companies_list` to see what it is. As the other comments for similar methods, I still do not know why the offset is generated randomly.

File: bot/seasons/evergreen/game.py [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+        # Get release date separately with checking
+        release_date = dt.utcfromtimestamp(data["first_release_date"]).date() if "first_release_date" in data else "?"
+
+        # Create Age Ratings value
+        rating = ", ".join(f"{AgeRatingCategories(age['category']).name} {AgeRatings(age['rating']).name}"
+                           for age in data["age_ratings"]) if "age_ratings" in data else "?"
+
+        companies = ", ".join(comp["company"]["name"] for comp in data["involved_companies"]) \
+            if "involved_companies" in data else "?"
+
```
Comment (eivl) [inline bot/seasons/evergreen/game.py]: I do not like the way the multi-line has been handled. they are inconsistent and overly messy. 
I do not have a suggestion atm, but we can have a discussion about it at your leisure. The contribution guide might have a suggestion about what to do.

File: bot/seasons/evergreen/game.py [hunk 25]
```diff
@@ -0,0 +1,374 @@
...
+        """Create good formatted Game Company page."""
+        # Generate URL of company logo
+        url = LOGO_URL.substitute({"image_id": data["logo"]["image_id"] if "logo" in data else ""})
+
+        # Try to get found date of company
+        founded = dt.utcfromtimestamp(data["start_date"]).date() if "start_date" in data else "?"
+
+        # Generate list of games, that company have developed or published
+        developed = ", ".join(game["name"] for game in data["developed"]) if "developed" in data else "?"
+        published = ", ".join(game["name"] for game in data["published"]) if "published" in data else "?"
```
Comment (eivl) [inline bot/seasons/evergreen/game.py]: this is a great example of a readable, well-written comprehension and ternary expression.

File: bot/seasons/evergreen/game.py [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+        if not 1 <= amount <= 25:
+            await ctx.send("Your provided amount is out of range. Our minimum is 1 and maximum 25.")
+            return
+
+        # Get games listing, if genre don't exist, show error message with possibilities.
+        # Offset must be random, due otherwise we will get always same result (offset show in which position should
+        # API start returning result)
+        try:
+            games = await self.get_games_list(amount, self.genres[genre],
+                                              offset=random.randint(0, 150))
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]:
```suggestion
            games = await self.get_games_list(amount, self.genres[genre], offset=random.randint(0, 150))
```

File: bot/seasons/evergreen/game.py [hunk 27]
```diff
@@ -0,0 +1,382 @@
...
+
+        self.genres = {}
+
+        # Replace complex names with names from ALIASES
+        for genre in genres:
+            if genre in ALIASES:
+                for alias in ALIASES[genre]:
+                    self.genres[alias] = genres[genre]
+            else:
+                self.genres[genre] = genres[genre]
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: Since we only do `_get_genres()` **once**, if this fails we have to reload the entire cog everytime, so I suggest turning `self.genres` into a `property` instead like so
```py
@property
def gernes(self):
    ...
```
This way, we can also add another command to refresh `genres` whenever we want to.

Or, you can create a `discord.ext.tasks` that loop every x amount that will fetch the genres, and of course, a command to refresh `self.genres`.

Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: since `genres` is a dictionary, you can loop over it's `.items()` to avoid having to do `genres[genre]` every time, something like
```py
for genre_name, genre in genres.items():
    ....
        self.genres[alias] = genre
```

File: bot/seasons/evergreen/game.py [hunk 28]
```diff
@@ -0,0 +1,403 @@
...
+            await ctx.send("Your provided amount is out of range. Our minimum is 1 and maximum 25.")
+            return
+
+        # Get games listing, if genre don't exist, show error message with possibilities.
+        # Offset must be random, due otherwise we will get always same result (offset show in which position should
+        # API start returning result)
+        try:
+            games = await self.get_games_list(amount, self.genres[genre], offset=random.randint(0, 150))
+        except KeyError:
+            possibilities = "`, `".join(difflib.get_close_matches(genre, self.genres))
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: Right now `difflib.get_close_matches()` is having inconsistent results, for example it cannot find `hack` and `slash` in `hack-and-slashes` so maybe you need a different way to suggest it?

image

image

File: bot/seasons/evergreen/game.py [hunk 29]
```diff
@@ -0,0 +1,403 @@
...
+            await ctx.send(f"Invalid genre `{genre}`. {f'Maybe you meant `{possibilities}`?' if possibilities else ''}")
+            return
+
+        # Create pages and paginate
+        pages = [await self.create_page(game) for game in games]
+
+        await ImagePaginator.paginate(pages, ctx, Embed(title=f"Random {genre} Games"))
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: Right now there are way too many blank lines between games, consider using the keyword `empty`

image

Thanks @sco1 for finding out about that!

**Comment by ikuyarihS** [comment]
Also right now it is impossible to find genres with spaces between them, consider using `Greedy` or parse the arguments manually instead

image

File: bot/seasons/evergreen/game.py (hunk 28 above)
Comment (eivl) [reply bot/seasons/evergreen/game.py]: interesting... what would fuzzy-matching return in this case?

**Comment by ks129** [comment]
@ikuyarihS I don't added auto recalling fetch for games when 1 item in possibilities, due this have conflict with `cutoff`.

File: bot/seasons/evergreen/game.py [hunk 30]
```diff
@@ -0,0 +1,395 @@
...
+            await ctx.send("Your provided amount is out of range. Our minimum is 1 and maximum 25.")
+            return
+
+        # Get games listing, if genre don't exist, show error message with possibilities.
+        # Offset must be random, due otherwise we will get always same result (offset show in which position should
+        # API start returning result)
+        try:
+            games = await self.get_games_list(amount, self.genres[genre], offset=random.randint(0, 150))
+        except KeyError:
+            possibilities = "`, `".join(difflib.get_close_matches(genre, self.genres, cutoff=0.4))
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: WIth `cutoff=0.4` the results seems a bit wild, so maybe increasing this back?
image

File: bot/seasons/evergreen/game.py [hunk 31]
```diff
@@ -0,0 +1,395 @@
...
+    async def genres(self, ctx: Context) -> None:
+        """Get all available genres."""
+        await ctx.send(f"Currently available genres: {', '.join(f'`{genre}`' for genre in self.genres)}")
+
+    @games.command(name="search", aliases=["s"])
+    async def search(self, ctx: Context, *, search_term: str) -> None:
+        """Find games by name."""
+        lines = await self.search_games(search_term)
+
+        await LinePaginator.paginate(lines, ctx, Embed(title=f"Game Search Results: {search_term}"), empty=False)
```
Comment (ikuyarihS) [inline bot/seasons/evergreen/game.py]: This is a minor issue, can be ignored and addressed later - right now there is no space between each results, so either move the rating up to the name or add a space in between.

image
